### ISSUE TITLE
Spotify connect not playing over bluetooth like other audio sources

### ISSUE THREAD

**Issue by roycenich** [balbuze/volumio-plugins#158]
Hello, I installed this popular plugin to enable bluetooth headphones:

https://volumio.org/forum/bluetooth-speaker-plugin-t7432.html

and it works great for any stream except for Spotify. I only hear Spotify through the headphone jack.

This isn't really an issue with your plugin per say, but I was thinking maybe you know why the audio output to a bluetooth sink might be different than say internet radio?

**Comment by ashthespy** [comment]
The default output device is determined from Volumio's Playback settings via `commandRouter.sharedVars.get('alsa.outputdevice')`. I am not sure if the `bluetooth-speaker-plugin` updates this.  
But if you know what the bluetooth output device is - you can manually change it in `startconnect.sh`, but any new changes in the plugin settings page will rewrite this.
The right way would be for the `bluetooth-speaker-plugin` to change the default volumio output device, and let volumio know that it has changed it!

**Comment by roycenich** [comment]
Thanks for the help! This is the plugin script below.

Maybe you can help me understand what to set the backend to in the `startconnect.sh` file?

```
#!/bin/bash

echo "Installing Bluetooth Dependencies"
sudo apt-get update
sudo apt-get install -y libasound2-dev dh-autoreconf libortp-dev bluez pi-bluetooth bluez-tools libbluetooth-dev libusb-dev libglib2.0-dev libudev-dev libical-dev libreadline-dev libsbc1 libsbc-dev

echo "Cloning Bluez-Alsa repo"
cd /tmp
git clone https://github.com/Arkq/bluez-alsa.git

echo "Building Bluez-Alsa"
cd bluez-alsa
autoreconf --install
mkdir build && cd build
../configure --disable-hcitop --with-alsaplugindir=/usr/lib/arm-linux-gnueabihf/alsa-lib
make 

echo "Installing Bluez-Alsa"
sudo make install

cat > /lib/systemd/system/bluezalsa.service <<EOC
[Unit]
Description=BluezAlsa proxy
Requires=bluetooth.service
After=bluetooth.service
[Service]
Type=simple
User=root
Group=audio
ExecStart=/usr/bin/bluealsa --disable-hfp
[Install]
WantedBy=multi-user.target
EOC

sudo systemctl daemon-reload
sudo systemctl enable bluezalsa.service

#requred to end the plugin install
echo "plugininstallend"
```

**Comment by ashthespy** [comment]
That doesn't provide the info - but from the [bluez-alsa configuration guide](https://github.com/Arkq/bluez-alsa#configuration--usage) I'd reckon you might find with `cat ~/.asoundrc`.
I couldn't find the git of bluetooth-speaker-plugin, else I could have helped more..

**Comment by balbuze** [comment]
I had a quick look at the bluetooth plugin. It seems to not configure the volumio output, but only mpd config. Two way : installing the last dev version of volspotconnect2 ad modify bye hand startconnect.sh or contact the author of the bluetooth plugin in order to make it more "standard"

**Comment by balbuze** [comment]
hi@ash!

**Comment by balbuze** [comment]
The last version dev of volspotconnect2 only update startconnect when you save setting. So no risk of auto update...

**Comment by ashthespy** [comment]
>I had a quick look at the bluetooth plugin. It seems to not configure the volumio output, but only mpd config. Two way : installing the last dev version of volspotconnect2 ad modify bye hand startconnect.sh or contact the author of the bluetooth plugin in order to make it more "standard"

@balbuze  I completely agree with this, that is motivation of using a complete system like Volumio, and not manually installing mpd + librespot + gpio+ lcd + blah + blah ;-) 

@roycenich  I'd urge you to open up an issue in the authors git about changing the default source. I think most of balbuze's `audio_interface`  plugins do this already, if I am not mistaken. 

>The last version dev of volspotconnect2 only update startconnect when you save setting. So no risk of auto update..

Ah, neat!

**Comment by roycenich** [comment]
Thanks @ashthespy, I looked at the git and noticed there is code commented out that might do it the "correct way" so I will try and uncomment that out and see what happens!

In case that fails, can you tell what I would manually to `startconnect.sh` in order to select the correct audio output?  

> /*
> BluetoothController.prototype.writeAsoundFile = function(mac) {
>     var self = this;
>     var defer = libQ.defer();
>     self.logger.info('Change softmixer device for audio device to:' + mac);
> 
> 
>     var asoundcontent = '';
> 
>     if (mac !== undefined)
>     {
>         asoundcontent += 'defaults.bluealsa { \n';
>         asoundcontent += 'interface "hci0"            # host Bluetooth adapter \n';
>         asoundcontent += '   device "' + mac + '"  # Bluetooth headset MAC address \n';
>         asoundcontent += '   profile "a2dp" \n';
>         asoundcontent += '}\n';
>     }
> 
> 
>     fs.writeFile('/home/volumio/.asoundrc', asoundcontent, 'utf8', function(err) {
>         if (err) {
>             self.logger.info('Cannot write /var/lib/mpd/.asoundrc: ' + err);
>         } else {
>             self.logger.info('asoundrc file written');
>             var mv = execSync('/usr/bin/sudo /bin/mv /home/volumio/.asoundrc /var/lib/mpd/.asoundrc', { uid:1000, gid: 1000, encoding: 'utf8' });
>             var apply = execSync('/usr/sbin/alsactl -L -R nrestore', { uid:1000, gid: 1000, encoding: 'utf8' });
>             var apply3 = execSync('/usr/sbin/alsactl -L -R nrestore', { uid:1000, gid: 1000, encoding: 'utf8' });
>         }
>     });
> };
> 
> 
> */ 

I did post and issue in his git as I realize this is no fault of this plugin. You guys are more responsive, so thanks for helping. I'm a noob with these settings but I feel so close.

**Comment by ashthespy** [comment]
Good luck! :-) 

For the `startconnect.sh` story - just change the parameter for the device flag, for eg  `--device bluealsa` and then restart the daemon with `sudo systemctl restart volspotconnect2.service`

**Comment by roycenich** [comment]
Okay so I got volspotconnect2 working by changing the --device setting in

`/data/plugins/music_service/volspotconnect2/startconnect.sh`

to: 

`bluealsa:HCI=hci0,DEV=E9:09:EF:A6:29:D6,PROFILE=a2dp`

And I also had to change my `~/.asoundrc` to read:


```
pcm.!default {
    type plug
    slave.pcm {
        type bluealsa
        interface "hci0"
        device "E9:09:EF:A6:29:D6"
        profile "a2dp"
    }
}
ctl.!default {
    type bluealsa
    interface "hci0"
}
```

And it plays great. Doesn't even skip or cut out when on wifi. 

I haven't been able to change the device settings for the Spotify plugin though. I tried changing the `output_name` in the /etc/spopd.conf but it didn't help. If you know what to change for that file please let me know. I have no idea what sox is:


```
[spop]
spotify_username = divinevette
spotify_password = ScorPion84


high_bitrate = false
audio_output = sox
pretty_json = true
search_results = 50
cache_path =/run/shm

[sox]
output_type = alsa
output_name = hw:0
```

**Comment by ashthespy** [comment]
@roycenich change your spotify password!

**Comment by balbuze** [comment]
I guess  it's already a fake password... devinette means riddle... but it is just my opinion...

**Comment by roycenich** [comment]
So embarrassed! Thanks for the help!
